Drawing function samples

rossviljoen opened the pull request:
Closes #22 

This PR implements the [Random API ](https://docs.julialang.org/en/v1/stdlib/Random/#An-optimized-sampler-with-pre-computed-data) for drawing random function samples from a BLR.

e.g. if you have `f = BayesianLinearRegressor(mw, Λw)`, calling `gs = rand(f, n)` will give you an array of `n` functions which can be evaluated at any inputs `X`.

@willtebbutt

rossviljoen commented on src/bayesian_linear_regression.jl:
```diff
@@ -73,3 +73,25 @@ function __compute_inference_quantities(fx::FiniteBLR, y::AbstractVector{<:Real}
+    w::Tw
+end
+
+(s::BLRFunctionSample)(X::AbstractMatrix{<:Real}) = X's.w
+(s::BLRFunctionSample)(X::ColVecs) = X.X's.w
+(s::BLRFunctionSample)(X::RowVecs) = X.X * s.w
```
Since the package assumes that `X <: AbstractMatrix` - are the ColVecs/RowVecs methods needed?

rossviljoen replied:
Actually - might be worth it to make the entire package work with either ColVecs/RowVecs?

willtebbutt replied:
I don't actually recall _why_ I'm using matrices all over the place in this package.... definitely something we should change for the sake of consistency.

rossviljoen replied:
#25

rossviljoen commented on src/bayesian_linear_regression.jl:
```diff
@@ -73,3 +73,25 @@ function __compute_inference_quantities(fx::FiniteBLR, y::AbstractVector{<:Real}
+
+function Random.Sampler(::Type{<:AbstractRNG}, blr::BayesianLinearRegressor, ::Random.Repetition)
+    Λw_U = _cholesky(blr.Λw).U
+    Random.SamplerSimple(blr, (Λw_U=Λw_U, mw=blr.mw))
```
I'm just using this sampler to precompute the cholesky of the precision matrix - but since it's used in pretty much everything anyway - probably worth it to add the cholesky to the BLR struct directly?

rossviljoen replied:
Or use an `AbstractMvNormal` for w instead :^)

rossviljoen replied:
#26

willtebbutt commented on test/bayesian_linear_regression.jl:
```diff
@@ -121,4 +121,24 @@ end
+
+        # test statistical properties of the sampled functions
+        gs = rand(rng, f, samples)
+        Y = hcat(map(h -> h(X), gs)...)
```
```suggestion
        Y = reduce(hcat, map(h -> h(X), gs))
```
generally much kinder on the compiler because splatting is only really designed for a few arguments, or something like that. Either way, there's a specialised method of `reduce` for both `hcat` and `vcat` which are generally a good choice if the number of things you're concatenating is really large.

rossviljoen replied:
thanks - I did think that was a horrendous way of doing it :D

willtebbutt reviewed (commented):
This is a really nice PR, thanks for opening. Just a few things to address before we merge.

rossviljoen commented on src/bayesian_linear_regression.jl:
```diff
@@ -73,3 +73,25 @@ function __compute_inference_quantities(fx::FiniteBLR, y::AbstractVector{<:Real}
+    Random.SamplerSimple(blr, (Λw_U=Λw_U, mw=blr.mw))
+end
+
+function Random.rand(rng::AbstractRNG, sp::Random.SamplerSimple{<:BayesianLinearRegressor})
```
Ah, yes should be fairly easy

rossviljoen replied:
Could also add a way to sample a single function with N ouputs (instead of a vector of N functions as currently) - would be much faster for large N for the same reason as this

willtebbutt replied:
> Ah, yes should be fairly easy

Ace.

> Could also add a way to sample a single function with N ouputs (instead of a vector of N functions as currently) - would be much faster for large N for the same reason as this

What kind of thing do you have in mind here? Are you talking about multi-output BLR, or just evaluating multiple samples from the same BLR quickly?

rossviljoen replied:
I mean evaluating multiple samples from a single-output BLR more efficiently

rossviljoen replied:
> Ah, yes should be fairly easy

oh, it's actually not that easy because the multisample version can ask for an arbitrary number of dimensions

willtebbutt replied:
> oh, it's actually not that easy because the multisample version can ask for an arbitrary number of dimensions

Oh, really? Could you elaborate?

rossviljoen replied:
added the multisample version

rossviljoen commented:
@willtebbutt happy with this now?

willtebbutt reviewed (approved):
Ah, I was imagining implementing an out-of-place multi-sample version. I would quite like to be able to AD-through multi-sample generation 😬 .

I'm going to approve now, but I would appreciate it if you would open an issue regarding multi-sample generation so that we don't forget.

Also, patch bump please :)

willtebbutt commented on test/bayesian_linear_regression.jl:
```diff
@@ -121,4 +121,27 @@ end
+        @test g(X) == g(Xr)
+
+        samples1, samples2 = 10_000, 1000
+        gs = rand(rng, f, samples1, samples2)
```
Actually, scratch the above, I've realised I don't understand what's going on here. What should I expect `gs` to be?

willtebbutt replied:
Oh, wait, nvm. It produces a matrix of samples. I had literally no idea that was a thing that the distributions API permitted. Or this a std lib thing?

rossviljoen replied:
`g = rand(rng, f)` returns a `BLRFunctionSample`
`gs = rand(rng, f, n, m, o)` returns an array of `BLRFunctionSample` where `size(gs) = (n, m, o)`

rossviljoen replied:
That's the convention of the `Random` API which I tried to follow

rossviljoen replied:
But I suppose since we need non-mutating rand using the `Sampler` interface is a bit pointless anyway

willtebbutt replied:
> But I suppose since we need non-mutating rand using the Sampler interface is a bit pointless anyway

It's not obvious to me that this is true. In particular, the single-sample version of `rand` that you've implemented appears to me to be mutation-free. Is that not the case?

rossviljoen commented:
@willtebbutt Changed it to non-mutating and added some autodiff tests (just copying what's here already)

willtebbutt commented:
Looks like lines 87 and 89 aren't covered by the tests. Are they redundant now that you're not using the sampler interface?

rossviljoen commented:
I think it can be useful to have `Random.Sampler` defined anyway (e.g. https://github.com/JuliaStats/Distributions.jl/pull/1314)

But I removed `Random.eltype` since that's definitely not needed anymore.

Also added back the in-place rand because why not

willtebbutt commented:
I have literally no idea why the ubuntu tests are failing. Any thoughts @rossviljoen ?

Also, could you bump the patch?

rossviljoen commented:
Absolutely no clue about ubuntu - tests run fine locally on arch & julia 1.7 so I'm pretty sure it's unrelated. Looks like the CI needs updating anyway.

Bumped the patch - so should I just merge or open another PR to fix CI first?

willtebbutt commented:
Please merge and tag. If you've got time to sort the CI out, that would be greatly appreciated.
